### ISSUE TITLE
Value incorrectly flagged as being possibly None after `or` predicate ensures it isn't None

### ISSUE THREAD

**Issue by rowillia** [python/mypy#3508]
Invoked with: ` python3.6 -m mypy --warn-no-return --strict-optional test_or_shortcut.py`

```python
from typing import Optional, Any


def test_or_shortcut(value: Optional[Any]) -> None:
    if not value:
        pass
    if not value or value.get('foo') == 'hello':
        pass
```
Fails with:
```
test_or_shortcut.py:7: error: Item "None" of "Optional[Any]" has no attribute "get"
```

Oddly if I remove the first branch things typecheck fine:

```python
from typing import Optional, Any


def test_or_shortcut(value: Optional[Any]) -> None:
    if not value or value.get('foo') == 'hello':
        pass
```

**Comment by gvanrossum** [comment]
Hm, that looks like a binder bug. But I can't repro it. Does it still fail this way in master?

**Comment by rowillia** [comment]
@gvanrossum Repro'd on HEAD:

```bash
rwilliams at rwilliams-mbp114-2 in ~/src/oss/mypy (master●) (venv)
$ python3.6 -m mypy -V
mypy 0.520-dev-72168faf682722ff8ffa6876d3c2fb94a30d5012
```

**Comment by gvanrossum** [comment]
Ah, I did'nt copy the flags you are passing. Sorry. Labeled as bug affecting strict-optional.

**Comment by ddfisher** [comment]
This looks familiar to me -- I think this might be to do with `Any`s behavior in the subtyping check (or maybe with `is_subtype` being used instead of `is_proper_subtype`).

**Comment by ddfisher** [comment]
I was wrong -- it's actually related to truthiness tracking.  For some reason, after the `if not value: pass`, the `Any` part of the union has `can_be_true` set to `False`.

**Comment by gvanrossum** [comment]
What if there is something other than `pass` in the first `if` block?

**Comment by ddfisher** [comment]
I was wondering that too.  I tried replacing the `pass` with an unrelated assignment statement -- the problem persisted.

**Comment by ddfisher** [comment]
Okay, I think I've tracked this down to the root cause of the problem: `true_or_false` (in `types.py`) doesn't behave correctly on `Union`s.  In particular, it sets `can_be_true` and `can_be_false` to `True` on the *`Union` type*, but not on any of its items!

**Comment by lk-geimfari** [comment]
I have the same problem, but with `enums`

Here is my `enum` object:
```python
from enum import Enum
from random import choice

class PrefixSign(Enum):
    POSITIVE = 'positive'
    NEGATIVE = 'negative'

    def get_random_item(self):
        return choice(list(self))
```
And when i tried this:
```python
def prefix(sign: Optional[PrefixSign] = None) -> str:
    prefixes = {
        'negative': ['...'],
        'positive': ['...'],
    }

    if sign is None:
        sign = PrefixSign.get_random_item()

    if sign in PrefixSign:
        return self.random.choice(prefixes[sign.value])
    else:
        raise NonEnumerableError(PrefixSign)
```

I got this:

```
project/test.py:24: error: Item "None" of "Optional[PrefixSign]" has no attribute "value"
project/units.py:25: error: Item "None" of "Optional[PrefixSign]" has no attribute "value"
```

How to deal with this?

**Comment by ilevkivskyi** [comment]
@lk-geimfari I think your particular case may be related to the fact that the method ``get_random_item()`` is not annotated. If I am right, then this should be fixed by the last commit. Could you please try your code with the current mypy master?

**Comment by lk-geimfari** [comment]
@ilevkivskyi Unfortunately, this did not help. I have updated the code which i write, please look above. The problem that i don't know which subclass of Enum will be returned and when i try this:

```python
def get_random_item(self) -> Enum:
    return choice(list(self))
```
I got even more errors:

```
project/utils.py:189: error: Incompatible return value type (got "Enum", expected "Gender")
project/providers/units.py:322: error: Incompatible types in assignment (expression has type "Enum", variable has type "Optional[PrefixSign]")
project/providers/units.py:325: error: Item "None" of "Optional[PrefixSign]" has no attribute "value"
project/providers/code.py:58: error: Incompatible types in assignment (expression has type "Enum", variable has type "Optional[ISBNFormat]")
```

**Comment by ilevkivskyi** [comment]
@lk-geimfari You can try self-types instead of just returning ``Enum``.
